src/ape/api/providers.py

```diff
@@ -1316,6 +1382,96 @@ def _find_txn_by_account_and_nonce(
                     stop_block,
                 )
 
+    def poll_blocks(
+        self,
+        stop_block: Optional[int] = None,
+        required_confirmations: Optional[int] = None,
+        new_block_timeout: Optional[int] = None,
+    ) -> Iterator[BlockAPI]:
+        filter = self.web3.eth.filter("latest")
+        network_name = self.network.name
+        block_time = self.network.block_time
+        timeout = (
+            (
+                10.0
+                if network_name == LOCAL_NETWORK_NAME or network_name.endswith("-fork")
+                else 50 * block_time
+            )
+            if new_block_timeout is None
+            else new_block_timeout
+        )
+
+        if required_confirmations is None:
+            required_confirmations = self.network_confirmations
+
+        last_yielded_height = None
+
+        while True:
+            if (
+                stop_block is not None
+                and last_yielded_height is not None
+                and last_yielded_height >= stop_block
+            ):
+                break
+            changes = self.web3.eth.get_filter_changes(filter.filter_id)
+            for new_block_hash in changes:
+                block = self.web3.eth.get_block(new_block_hash)
+                confirmed_block_number = block.number - required_confirmations
```

Comment:
Are we able to adjust the filter to account for the confs? Right now, it is set to `"latest"`, but could we refactor such that we account for the confs at the filter level and then just yield from there.

I think that will prevent some unnecessary `get_block()` requests.

Reply:
I also explored ways of doing this. The options web3.py gives us are:

- 'latest'
- 'pending'
- 'earliest'
- 'safe'
- 'finalized'

it sounds like 'safe' or 'finalized' would be what we want. Unfortunately, web3.py doesn't expose any way of customizing what 'safe' or 'finalized' means. The docs also say:

> Though `"latest"` and `"safe"` block identifiers are not yet part of the specifications for `eth_newFilter`, they are supported by web3.py and may or may not yield expected results depending on the node being accessed.

we can maybe specify a block number and re-create a new filter each time we're going to fetch a new set of blocks. This would require us only pinging the node for the latest block number, subtracting our number of required confs, and fetching that old block. 

Currently, we poll each block twice: once when its the newest block, and once to get its number. This is because when we poll for blocks, we only get a blockhash back, not a block number. In order to get the block number, we fetch this block by its hash.

We should be able to use `eth_getBlockNumber` and reduce this double-fetching of blocks, but it is not compatible with our current block polling. Instead of being notified of a new block, we would loop and call `eth_getBlockNumber` until the number increases.

Reply:
> we can maybe specify a block number and re-create a new filter each time we're going to fetch a new set of blocks. This would require us only pinging the node for the latest block number, subtracting our number of required confs, and fetching that old block.

This is kinda what i was imagining

> We should be able to use eth_getBlockNumber and reduce this double-fetching of blocks, but it is not compatible with our current block polling. 

Keep in mind if you in Web3Provider, you are allowed to assume EVM like behavior and RPCs (provided they are standard enough)

Reply:
Implemented polling logic the way you suggested here